### PR TITLE
Split up arrow::array::builder module (#1843)

### LINKED ISSUE #1843: Split up arrow::array::builder module
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

The arrow::array::builder module is implemented as one immense builder.rs file, this has a number of drawbacks:

* Types have access to each others private fields, which can make invariants hard to reason about #1842 
* The file is incredibly unwieldy
* It is unclear what test coverage each of the builders have

**Describe the solution you'd like**

I would like to split out the various builders from `builder.rs` into their own submodules, e.g.

```
builder/mod.rs
builder/buffer_builder.rs
builder/boolean_builder.rs
...
```

The top-level `builder/mod.rs` can then re-export each of these with `pub use` to avoid introducing breaking changes.

**Describe alternatives you've considered**

We could not do this

### PR DESCRIPTION
**Pull request by DaltonModlin**
# Which issue does this PR close?

Closes #1843 

# Rationale for this change
 
Currently, arrow::array::builder contains 12 distinct builders; breaking the builders into submodules makes test coverage and functional differences more obvious.

# Are there any user-facing changes?
No

### CODE REVIEW

File: arrow/src/array/builder/mod.rs [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+// under the License.
+
+//! Defines a [`BufferBuilder`](crate::array::BufferBuilder) capable
+//! of creating a [`Buffer`](crate::buffer::Buffer) which can be used
+//! as an internal buffer in an [`ArrayData`](crate::array::ArrayData)
+//! object.
+
+pub mod boolean_buffer_builder;
+pub mod boolean_builder;
+#[allow(clippy::module_inception)]
```
Comment (tustvold) [inline arrow/src/array/builder/mod.rs:25]: Why is this needed?

File: arrow/src/array/builder/mod.rs [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! Defines a [`BufferBuilder`](crate::array::BufferBuilder) capable
+//! of creating a [`Buffer`](crate::buffer::Buffer) which can be used
+//! as an internal buffer in an [`ArrayData`](crate::array::ArrayData)
+//! object.
+
+pub mod boolean_buffer_builder;
```
Comment (tustvold) [inline arrow/src/array/builder/mod.rs:23]: I think these can be private as they are re-exported?

**Review by tustvold — COMMENTED** [review]
Thank you for this 👍, I plan to go through this this evening unless someone else gets there first. Left some minor comments from a cursory run through.

Edit: CI appears to be unhappy

File: arrow/src/array/builder/mod.rs (hunk 1 above)
Comment (DaltonModlin) [reply arrow/src/array/builder/mod.rs:25]: Module inception isn't necessary anymore. I originally copy-pasted the dependency list, and this was part of it. I'll go ahead and remove it.

**Review by alamb — APPROVED** [review]
Thank you very much @DaltonModlin 

this PR appears to have a conflict -- could you possible resolve it and I will then merge it in?

Thanks!

**Comment by alamb** [comment]
Thanks again @DaltonModlin
